### DISCUSSION TITLE
Exception thrown while attempting to evaluate the LINQ query parameter expression

### DISCUSSION THREAD

**Discussion by MarcoGaribaldi** [Finbuckle/Finbuckle.MultiTenant#792] [Q&A] [answered]
Hello everyone,

hope someone can help me with the following problem. I am working on an app where the backend is based on the [dotnet-webapi-starter-kit](https://github.com/fullstackhero/dotnet-webapi-starter-kit). This template has a db seeder. When the application starts and the ApplicationDbSeeder class is executed, I get an error in the SeedRolesAsync method when the following code is executed

	private async Task SeedRolesAsync(ApplicationDbContext dbContext)
	{
		foreach (string roleName in Roles.DefaultRoles)
		{
			if (await roleManager.Roles.SingleOrDefaultAsync(r => r.Name == roleName)  <=== Exception occurs here
				is not { } role)
			{
				etc...
			
I get this error:

"An exception was thrown while attempting to evaluate the LINQ query parameter expression 'value(Finbuckle.MultiTenant.EntityFrameworkCore.FinbuckleEntityTypeBuilderExtensions+ExpressionVariableScope).Context.TenantInfo'. See the inner exception for more information."

The inner exception is:

"Object reference not set to an instance of an object."

Finbuckle is configured this way:

	internal static IServiceCollection AddMultitenancy(this IServiceCollection services, IConfiguration config)
	{
		return services
			.AddDbContext<TenantDbContext>((p, m) =>
			{
				var databaseSettings = p.GetRequiredService<IOptions<DatabaseSettings>>().Value;
				m.UseDatabase(databaseSettings.DBProvider, databaseSettings.ConnectionString);
			})
			.AddMultiTenant<FSHTenantInfo>()
				.WithClaimStrategy(FSHClaims.Tenant)
				.WithHeaderStrategy(MultitenancyConstants.TenantIdName)
				.WithQueryStringStrategy(MultitenancyConstants.TenantIdName)
				.WithEFCoreStore<TenantDbContext, FSHTenantInfo>()
				.Services
			.AddScoped<ITenantService, TenantService>();
	}

I have noticed that the same error happens whenever LINQ code is executed. As an example, SeedRolesAsync is called by the following method

	public async Task SeedDatabaseAsync(ApplicationDbContext dbContext, CancellationToken cancellationToken)
	{
		await SeedRolesAsync(dbContext);
		await SeedAdminUserAsync();
		await seederRunner.RunSeedersAsync(cancellationToken);
	}

If I comment out SeedRolesAsync the same error happens in the SeedAdminUserAsync method when this line is executed

    if (await userManager.Users.FirstOrDefaultAsync(u => u.Email == currentTenant.AdminEmail)
        is not { } adminUser)
		
The same happens if I do the following

    if (!userManager.Users.Any())
        return;
		
Any idea on what may be the problem?

**Comment by AndrewTriesToCode** [comment, marked as answer]
Hi, so the root problem is that it doesn’t know the tenant at the time the db context was created. 

Typically in seeding it is not part of an actual request or user action so there is no tenant that makes sense. is the data you want to seed per tenant or global?

For the former, one option I’ve used successfully is to loop through my tenants and create a standalone db context for the current tenant and deed it as needed. Also I reused this logic for any new tenant created to seed its data at the time the tenant was registered.

For the latter, I’ve used a dummy or default tenant that allows me to get the db context to set non per tenant data as needed but isn’t a real tenant/customer.

**Comment by Godiex** [comment]
I have the same ploblem, i dont know how resolve the proble of the current tenant

**Comment by Maofaa-CIT** [comment]
I had the same problem and used the following solution:
if (await userManager.Users.IgnoreQueryFilters().FirstOrDefaultAsync(u => u.Email == currentTenant.AdminEmail)
is not { } adminUser)
I don't know if it will work with Finbuckle.MultiTenant or not.

**Comment by Maofaa-CIT** [comment]
hi @MarcoGaribaldi 
To solve this problem **Disable automatic index creation for RoleName**

protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    base.OnModelCreating(modelBuilder);
    
  
    modelBuilder.Entity<IdentityRole>(entity =>
    {
        entity.HasIndex(r => r.NormalizedName)
            .HasDatabaseName("RoleNameIndex")
            .IsUnique(false); 
    });
}
